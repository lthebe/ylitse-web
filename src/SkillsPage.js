import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { withStyles } from 'material-ui/styles';
import Typography from 'material-ui/Typography';
import CloseIcon from 'material-ui-icons/Close';
import Snackbar from 'material-ui/Snackbar';
import GridList from 'material-ui/GridList';
import IconButton from 'material-ui/IconButton';
import TextField from 'material-ui/TextField';
import SkillsListItem from './SkillsListItem';
import Page from './Page';

const styles = theme => ({
    row: {
        marginTop: theme.spacing.unit * 2,
        marginRight: theme.spacing.unit,
    },
    root: {
        display: 'flex',
        flexWrap: 'wrap',
        justifyContent: 'space-around',
        overflow: 'hidden',
        backgroundColor: theme.palette.background.paper,
    },
    gridList: {
        dislay: 'flex',
        flexWrap: 'nowrap',
        overflowX: 'auto',
    },
    titleBar: {
        background:
      'linear-gradient(to bottom, rgba(0,0,0,0.7) 0%, ' +
      'rgba(0,0,0,0.3) 70%, rgba(0,0,0,0) 100%)',
    },
});

class SkillsPage extends Component {
    static propTypes = {
        classes: PropTypes.shape({
            root: PropTypes.string,
            gridList: PropTypes.string,
            titleBar: PropTypes.string,
        }).isRequired,
    }
    constructor(props) {
        super(props);
        this.state = {
            errors: {},
            errorOpen: false,
            errorMessage: '',
            newSkill: '',
            skills: [],
        };
    }
    componentDidMount() {
        this.fetchSkills();
    }

    fetchSkills = async () => {
        try {
            // basically, all are same apiUrl, can I move somewhere?
            const apiUrl = process.env.API_URL || 'http://127.0.0.1:8080';
            const resp = await fetch(`${apiUrl}/skills`);
            const data = await resp.json();
            this.setState({ skills: data.resources });
        } catch (e) {
            this.openError(e.message);
        }
    };

    sendSkill = async (skill) => {
        const apiUrl = process.env.API_URL || 'http://127.0.0.1:8080';
        try {
            await fetch(`${apiUrl}/skills`, {
                headers: {
                    'Content-Type': 'application/json',
                },
                method: 'POST',
                body: JSON.stringify(skill),
            });
            this.fetchSkills();
        } catch (e) {
            this.openError(e.message);
        }
    }

    openError = (errorMessage) => {
        this.setState({ errorMessage, errorOpen: true });
    }

    closeError = () => {
        this.setState({ errorOpen: false });
    }

    deleteSkill = async (id) => {
        const apiUrl = process.env.API_URL || 'http://127.0.0.1:8080';
        try {
            await fetch(`${apiUrl}/skills/${id}`, {
                method: 'DELETE',
            });
            this.setState(prevState => ({
                skills: prevState.skills.filter(s => s.id !== id),
            }));
        } catch (e) {
            this.openError(e.message);
        }
    }

    addSkill = (event) => {
        switch (event.key) {
            case 'Enter': {
                if (this.state.errors.newSkill || !this.state.newSkill) {
                    return;
                }
                this.sendSkill({ name: this.state.newSkill });
                this.setState({
                    newSkill: '',
                });
                break;
            }
            case 'Escape': {
                this.setState(prevState => ({
                    newSkill: '',
                    errors: { ...prevState.errors, newSkill: '' },
                }));
                break;
            }
            default:
                break;
        }
    }

    updateValue = ({ target }) => {
        const { name, value } = target;
        const { errors } = this.state;
        let valid;
        switch (name) {
            case 'newSkill':
                valid = !(this.state.skills
                    .some(skill =>
                        (skill.name.toLowerCase() === value.toLowerCase())));
                errors.newSkill = valid ? '' : 'Already in list';
                break;
            default:
                break;
        }
        this.setState({
            [target.name]: target.value,
            errors,
        });
    }

    render() {
        const { classes } = this.props;
        return (
            <Page>
                <Typography variant="headline" component="h3">
                 Mentor skills
                </Typography>
                <TextField
                    name="newSkill"
                    label="Add a skill"
                    className={classes.row}
                    value={this.state.newSkill}
                    error={Boolean(this.state.errors.newSkill)}
                    helperText={this.state.errors.newSkill}
                    onChange={this.updateValue}
                    onKeyDown={this.addSkill}
                />

                <div className={classes.root}>
                    <GridList
                        cellHeight={60}
                        spacing={5}
                        className={classes.gridList}
                    >
                        {this.state.skills
                            .map(skill => (
                                <SkillsListItem
                                    key={skill.id}
                                    deleteSkill={() => {
                                        this.deleteSkill(skill.id);
                                    }}
                                    label={skill.name}
                                />
                            ))}
                    </GridList>
                </div>
                <Snackbar
                    open={this.state.errorOpen}
                    message={this.state.errorMessage}
                    autoHideDuration={3000}
                    action={
                        <IconButton color="inherit" onClick={this.closeError}>
                            <CloseIcon />
                        </IconButton>
                    }
                    onClose={this.closeError}
                />
            </Page>
        );
    }
}

export default withStyles(styles)(SkillsPage);
