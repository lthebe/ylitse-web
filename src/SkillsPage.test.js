import React from 'react';
import { shallow } from 'enzyme';

import SkillsPage from './SkillsPage';

describe('SkillsPage', () => {
    test('renders correctly', () => {
        const wrapper = shallow(<SkillsPage />).dive();

        expect(wrapper).toMatchSnapshot();
    });

    test('fetch error message is shown', async () => {
        const wrapper = shallow(<SkillsPage />).dive();
        const error = 'Can not fetch skills';

        window.fetch = jest.fn();
        window.fetch.mockReturnValue({
            json: () => {
                throw new Error(error);
            },
        });
        await wrapper.instance().fetchSkills();
        wrapper.update();
        expect(wrapper.state('feedbackOpen')).toBe(true);
        expect(wrapper.find({ message: error, open: true })).toHaveLength(1);

        wrapper.instance().closeFeedback();
        wrapper.update();
        expect(wrapper.state('feedbackOpen')).toBe(false);
        expect(wrapper.find({ message: error, open: false })).toHaveLength(1);
    });

    test('skills are fetched', async () => {
        const wrapper = shallow(<SkillsPage />).dive();
        const skills = [{ id: 23333, name: 'Cooking' }];
        const res = new Response(`{"resources": ${JSON.stringify(skills)}}`, {
            status: 200,
            headers: {
                'content-type': 'application/json',
            },
        });

        window.fetch = jest.fn();
        window.fetch.mockReturnValue(Promise.resolve(res));
        await wrapper.instance().fetchSkills();
        wrapper.update();
        expect(wrapper.state('skills')).toEqual(expect.arrayContaining(skills));
    });
});
