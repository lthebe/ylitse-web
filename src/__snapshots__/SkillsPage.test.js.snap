// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SkillsPage renders correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SkillsPage
    classes={
      Object {
        "input": "SkillsPage-input-1",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <WithStyles(Typography)
          component="h3"
          variant="headline"
        >
          Mentor skills
        </WithStyles(Typography)>,
        <TextField
          className="SkillsPage-input-1"
          error={false}
          helperText={undefined}
          label="Add a skill"
          name="newSkillName"
          onChange={[Function]}
          onKeyDown={[Function]}
          required={false}
          select={false}
          value=""
        />,
        <WithStyles(GridList)
          className="SkillsPage-input-1"
        />,
        undefined,
        <WithStyles(Snackbar)
          action={
            <WithStyles(IconButton)
              color="inherit"
              onClick={[Function]}
            >
              <pure(Close) />
            </WithStyles(IconButton)>
          }
          autoHideDuration={3000}
          message=""
          onClose={[Function]}
          open={false}
        />,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": "Mentor skills",
          "component": "h3",
          "variant": "headline",
        },
        "ref": null,
        "rendered": "Mentor skills",
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "className": "SkillsPage-input-1",
          "error": false,
          "helperText": undefined,
          "label": "Add a skill",
          "name": "newSkillName",
          "onChange": [Function],
          "onKeyDown": [Function],
          "required": false,
          "select": false,
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": Array [],
          "className": "SkillsPage-input-1",
        },
        "ref": null,
        "rendered": Array [],
        "type": [Function],
      },
      undefined,
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "action": <WithStyles(IconButton)
            color="inherit"
            onClick={[Function]}
          >
            <pure(Close) />
          </WithStyles(IconButton)>,
          "autoHideDuration": 3000,
          "message": "",
          "onClose": [Function],
          "open": false,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <WithStyles(Typography)
            component="h3"
            variant="headline"
          >
            Mentor skills
          </WithStyles(Typography)>,
          <TextField
            className="SkillsPage-input-1"
            error={false}
            helperText={undefined}
            label="Add a skill"
            name="newSkillName"
            onChange={[Function]}
            onKeyDown={[Function]}
            required={false}
            select={false}
            value=""
          />,
          <WithStyles(GridList)
            className="SkillsPage-input-1"
          />,
          undefined,
          <WithStyles(Snackbar)
            action={
              <WithStyles(IconButton)
                color="inherit"
                onClick={[Function]}
              >
                <pure(Close) />
              </WithStyles(IconButton)>
            }
            autoHideDuration={3000}
            message=""
            onClose={[Function]}
            open={false}
          />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": "Mentor skills",
            "component": "h3",
            "variant": "headline",
          },
          "ref": null,
          "rendered": "Mentor skills",
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "className": "SkillsPage-input-1",
            "error": false,
            "helperText": undefined,
            "label": "Add a skill",
            "name": "newSkillName",
            "onChange": [Function],
            "onKeyDown": [Function],
            "required": false,
            "select": false,
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [],
            "className": "SkillsPage-input-1",
          },
          "ref": null,
          "rendered": Array [],
          "type": [Function],
        },
        undefined,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "action": <WithStyles(IconButton)
              color="inherit"
              onClick={[Function]}
            >
              <pure(Close) />
            </WithStyles(IconButton)>,
            "autoHideDuration": 3000,
            "message": "",
            "onClose": [Function],
            "open": false,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
